runtime/compiler/control/CompilationThread.cpp

```diff
@@ -8777,81 +8763,114 @@ TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary *portLib, void *
+         // If this is a cold/warm compilation that takes too much memory
+         // add a hint to avoid queuing a forced upgrade
+         // Look only during startup to avoid creating too many hints.
+         // If SCC is larger we could store hints for more methods
+         //
+         if (TR::Compiler->vm.isVMInStartupPhase(jitConfig))
             {
-            TR_Hotness hotness = that->_methodBeingCompiled->_optimizationPlan->getOptLevel();
-            if (hotness <= cold)
+            if (scratchSegmentProvider.regionBytesAllocated() > TR::Options::_memExpensiveCompThreshold)
                {
-               sc->addHint(method, TR_HintLargeCompCPUC);
+               TR_Hotness hotness = that->_methodBeingCompiled->_optimizationPlan->getOptLevel();
+               if (hotness <= cold)
+                  {
+                  sc->addHint(method, TR_HintLargeMemoryMethodC);
+                  }
+               else if (hotness == warm)
+                  {
+                  sc->addHint(method, TR_HintLargeMemoryMethodW);
+                  }
                }
-            else if (hotness == warm)
+
+            if (that->getCompilationInfo()->getMethodQueueSize() - that->getQszWhenCompStarted() > TR::Options::_cpuExpensiveCompThreshold)
                {
-               sc->addHint(method, TR_HintLargeCompCPUW);
+               TR_Hotness hotness = that->_methodBeingCompiled->_optimizationPlan->getOptLevel();
+               if (hotness <= cold)
+                  {
+                  sc->addHint(method, TR_HintLargeCompCPUC);
+                  }
+               else if (hotness == warm)
+                  {
+                  sc->addHint(method, TR_HintLargeCompCPUW);
+                  }
                }
             }
          }
       }
+#if defined(J9VM_OPT_JITSERVER)
+   catch (const JITServer::StreamFailure &e)
+      {
+      // Stream failure here means it was produced by one of the calls to sc->addHint
+      // Fail the compilation here since attempting to finish it will result
+      // in another StreamFailure
+      that->_methodBeingCompiled->_compErrCode = compilationStreamFailure;
+      metaData = NULL; // indicate that the compilation has failed for postCompilationTasks
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "compThreadID=%d JITServer StreamFailure: %s", that->getCompThreadId(), e.what());
+      Trc_JITServerStreamFailure(vmThread, that->getCompThreadId(),  __FUNCTION__, compiler->signature(), compiler->getHotnessName(), e.what());
 
+      if (compiler && compiler->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
+         {
+         compiler->getOptions()->closeLogFileForClientOptions();
+         }
+
+      if (compiler)
+         {
+         compiler->~Compilation();
```

Comment:
Don't we have to 
```
         if (compiler->getKnownObjectTable())
            compiler->freeKnownObjectTable();
```
Otherwise, if `compiler` becomes NULL we lose our chances.

Reply:
Actually I think it's better if we don't destroy the compilation object here, but let the code flow and reach `postCompilationTasks()` where code caches and data caches need to be released:
```
   if (_compiler)
      {
      // Unreserve the code cache used for this compilation
      if (_compiler->cg() && _compiler->cg()->getCodeCache())
         {
#if defined(J9VM_OPT_JITSERVER)
         // For JITServer we should wipe this method from the code cache
         if (_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
            _compiler->cg()->getCodeCache()->resetCodeCache();
#endif /* defined(J9VM_OPT_JITSERVER) */
         _compiler->cg()->getCodeCache()->unreserve();
         _compiler->cg()->setCodeCache(0);
         }
      // Unreserve the data cache
      TR_DataCache *dataCache = (TR_DataCache*)_compiler->getReservedDataCache();
      if (dataCache)
         {
         if (entry->_tryCompilingAgain && entry->_compErrCode == compilationAotCacheFullReloFailure)
            {
            TR_ASSERT(canDoRelocatableCompile, "Only AOT compilations can fail with compilationAotCacheFullReloFailure. entry=%p\n", entry);
            _reservedDataCache = dataCache;
            }
         else
            {
            TR_DataCacheManager::getManager()->makeDataCacheAvailable(dataCache);
            _compiler->setReservedDataCache(NULL);
            }
         }
      }
```

Reply:
This makes sense. I just wasn't completely sure that simply setting `metaData` to `NULL` would indicate to `postCompilationTasks` that the compilation has not been successful. But if that's the case I agree.